Moving CuSparseMatrix/SparseMatrix out of CuVector/Vector

chenguoguo opened the pull request:
I named the function CopyToVec() in CuSparseMatrix to be consistent with that function in SparseMatrix. Tests passed.

danpovey commented:
OK, that change is good, but we can go further, and get rid of a bunch of unnecessary code.
Firstly, we don't need the OtherReal version of CopyToVec.  And (sorry), I think both of them should be changed to be called CopyElementsToVec, it's less ambiguous.
But also the special CUDA kernel is not needed.  You can use sizeof(MatrixElement<Real>) / sizeof(Real) as a fake stride, and &(elements[0].weight) as the start of the array, and make use of cublas_copy.  
I noticed that the functions CopyColFromVec and CopyColFromMat also use an unnecessary CUDA kernel-- they can also make use of cublas_copy.

chenguoguo commented:
I just pushed a commit, message below:

1. Changed CopyToVec in CuSparseMatrix/SparseMatrix to CopyElementsToVec,
and removed OtherReal version. Note that I didn't remove the OtherReal
version of CopyElementsToVec in SparseVector as that function is used by
other function;
2. Replaced unnecessary CUDA kernels used in CopyElementsToVec,
CopyColFromVec and CopyColFromMat with cublas_copy. Note,
cuda_copy_col_from_mat_df and cuda_copy_col_from_mat_fd were not replaced.

On Tue, Jan 5, 2016 at 12:13 PM, Daniel Povey <notifications@github.com>
wrote:

> OK, that change is good, but we can go further, and get rid of a bunch of
> unnecessary code.
> Firstly, we don't need the OtherReal version of CopyToVec. And (sorry), I
> think both of them should be changed to be called CopyElementsToVec, it's
> less ambiguous.
> But also the special CUDA kernel is not needed. You can use
> sizeof(MatrixElement) / sizeof(Real) as a fake stride, and
> &(elements[0].weight) as the start of the array, and make use of
> cublas_copy.
>
> I noticed that the functions CopyColFromVec and CopyColFromMat also use an
> unnecessary CUDA kernel-- they can also make use of cublas_copy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/434#issuecomment-169118508>.
>

danpovey commented on src/cudamatrix/cu-sparse-matrix.cc:
```diff
@@ -57,7 +57,8 @@ template <typename Real>
 Real CuSparseMatrix<Real>::Sum() const {
 #if HAVE_CUDA == 1
```
Please rebase against the latest master and test.  there should be lines here:
  if (NumElements() == 0)
    return 0.0;
Please test by logging into a GPU machine and going 'make test' in src/ - this will test more thoroughly.

danpovey commented on src/cudamatrix/cu-sparse-matrix.cc:
```diff
@@ -70,7 +71,8 @@ template <typename Real>
-    CuVector<Real> element_vec(*this);
+    CuVector<Real> element_vec(this->NumElements(), kUndefined);
+    this->CopyElementsToVec(&element_vec);
     return element_vec.Norm(2);
```
This code is not right, you need account for the fact that the matrix's size is different from the number of nonzero elements.  Otherwise it's not really the frobenius norm.

chenguoguo replied:
We are computing sqrt(sum_{i=1}^N (x_i^2)) as the Frobenius norm right? Why
would the number of elements matter? Perhaps I misunderstood
CuVector.Norm()?

Guoguo

On Tue, Jan 5, 2016 at 3:31 PM, Daniel Povey <notifications@github.com>
wrote:

> In src/cudamatrix/cu-sparse-matrix.cc
> <https://github.com/kaldi-asr/kaldi/pull/434#discussion_r48911401>:
>
> > @@ -70,7 +71,8 @@ template <typename Real>
> >  Real CuSparseMatrix<Real>::FrobeniusNorm() const {
> >  #if HAVE_CUDA == 1
> >    if (CuDevice::Instantiate().Enabled()) {
> > -    CuVector<Real> element_vec(*this);
> > +    CuVector<Real> element_vec(this->NumElements(), kUndefined);
> > +    this->CopyElementsToVec(&element_vec);
> >      return element_vec.Norm(2);
>
> This code is not right, you need account for the fact that the matrix's
> size is different from the number of nonzero elements. Otherwise it's not
> really the frobenius norm.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/434/files#r48911401>.
>

danpovey replied:
oh sorry, I forgot the definition.

On Tue, Jan 5, 2016 at 4:06 PM, Guoguo Chen <notifications@github.com>
wrote:

> In src/cudamatrix/cu-sparse-matrix.cc
> <https://github.com/kaldi-asr/kaldi/pull/434#discussion_r48913804>:
>
> > @@ -70,7 +71,8 @@ template <typename Real>
> >  Real CuSparseMatrix<Real>::FrobeniusNorm() const {
> >  #if HAVE_CUDA == 1
> >    if (CuDevice::Instantiate().Enabled()) {
> > -    CuVector<Real> element_vec(*this);
> > +    CuVector<Real> element_vec(this->NumElements(), kUndefined);
> > +    this->CopyElementsToVec(&element_vec);
> >      return element_vec.Norm(2);
>
> We are computing sqrt(sum_{i=1}^N (x_i^2)) as the Frobenius norm right?
> Why would the number of elements matter? Perhaps I misunderstood
> CuVector.Norm()? Guoguo
> … <#1521441390d26195_>
> On Tue, Jan 5, 2016 at 3:31 PM, Daniel Povey <notifications@github.com>
> wrote: In src/cudamatrix/cu-sparse-matrix.cc <#434 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/434#discussion_r48911401>>: > @@
> -70,7 +71,8 @@ template <typename Real> > Real
> CuSparseMatrix<Real>::FrobeniusNorm() const { > #if HAVE_CUDA == 1 > if
> (CuDevice::Instantiate().Enabled()) { > - CuVector<Real>
> element_vec(*this); > + CuVector<Real> element_vec(this->NumElements(),
> kUndefined); > + this->CopyElementsToVec(&element_vec); > return
> element_vec.Norm(2); This code is not right, you need account for the fact
> that the matrix's size is different from the number of nonzero elements.
> Otherwise it's not really the frobenius norm. — Reply to this email
> directly or view it on GitHub <
> https://github.com/kaldi-asr/kaldi/pull/434/files#r48911401>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/434/files#r48913804>.
>

chenguoguo replied:
OK I merged master and unit tests passed.

Guoguo

On Tue, Jan 5, 2016 at 4:08 PM, Daniel Povey <notifications@github.com>
wrote:

> In src/cudamatrix/cu-sparse-matrix.cc
> <https://github.com/kaldi-asr/kaldi/pull/434#discussion_r48913949>:
>
> > @@ -70,7 +71,8 @@ template <typename Real>
> >  Real CuSparseMatrix<Real>::FrobeniusNorm() const {
> >  #if HAVE_CUDA == 1
> >    if (CuDevice::Instantiate().Enabled()) {
> > -    CuVector<Real> element_vec(*this);
> > +    CuVector<Real> element_vec(this->NumElements(), kUndefined);
> > +    this->CopyElementsToVec(&element_vec);
> >      return element_vec.Norm(2);
>
> oh sorry, I forgot the definition.
> … <#152144356c736f1f_>
> On Tue, Jan 5, 2016 at 4:06 PM, Guoguo Chen <notifications@github.com>
> wrote: In src/cudamatrix/cu-sparse-matrix.cc <#434 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/434#discussion_r48913804>>: > @@
> -70,7 +71,8 @@ template <typename Real> > Real
> CuSparseMatrix<Real>::FrobeniusNorm() const { > #if HAVE_CUDA == 1 > if
> (CuDevice::Instantiate().Enabled()) { > - CuVector<Real>
> element_vec(*this); > + CuVector<Real> element_vec(this->NumElements(),
> kUndefined); > + this->CopyElementsToVec(&element_vec); > return
> element_vec.Norm(2); We are computing sqrt(sum_{i=1}^N (x_i^2)) as the
> Frobenius norm right? Why would the number of elements matter? Perhaps I
> misunderstood CuVector.Norm()? Guoguo … <#1521441390d26195_> On Tue, Jan 5,
> 2016 at 3:31 PM, Daniel Povey <notifications@github.com> wrote: In
> src/cudamatrix/cu-sparse-matrix.cc <#434
> <https://github.com/kaldi-asr/kaldi/pull/434> (comment) <#434 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/434#discussion_r48911401>>>: >
> @@ -70,7 +71,8 @@ template <typename Real> > Real
> CuSparseMatrix<Real>::FrobeniusNorm() const { > #if HAVE_CUDA == 1 > if
> (CuDevice::Instantiate().Enabled()) { > - CuVector<Real>
> element_vec(*this); > + CuVector<Real> element_vec(this->NumElements(),
> kUndefined); > + this->CopyElementsToVec(&element_vec); > return
> element_vec.Norm(2); This code is not right, you need account for the fact
> that the matrix's size is different from the number of nonzero elements.
> Otherwise it's not really the frobenius norm. — Reply to this email
> directly or view it on GitHub <
> https://github.com/kaldi-asr/kaldi/pull/434/files#r48911401>. — Reply to
> this email directly or view it on GitHub <
> https://github.com/kaldi-asr/kaldi/pull/434/files#r48913804>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/434/files#r48913949>.
>

danpovey commented:
Thanks.  In the longer term it would be better to create versions of cublas_copy with our own kernels, that work for different types, which is more generic than cuda_copy_col_from_mat_fd  and the like.  Could you please create an issue to track this?  I'll merge this now.

danpovey commented on src/cudamatrix/cu-sparse-matrix.cc:
```diff
@@ -202,6 +206,27 @@ void CuSparseMatrix<double>::CopyToSmat(SparseMatrix<float> *smat) const;
 void CuSparseMatrix<double>::CopyToSmat(SparseMatrix<double> *smat) const;
 
+template <typename Real>
+void CuSparseMatrix<Real>::CopyElementsToVec(CuVector<Real> *vec) const {
```
Oops.  I just noticed, this should be CuVectorBase, not CuVector.

danpovey commented on src/cudamatrix/cu-sparse-matrix.h:
```diff
@@ -95,6 +95,10 @@ class CuSparseMatrix {
 
+  /// Copy elements to CuVector. It is the caller's responsibility to resize
+  /// <*vec>.
+  void CopyElementsToVec(CuVector<Real> *vec) const;
```
should be CuVectorBase

chenguoguo replied:
That should be changed yes, I fixed that.

Also created an issue for cublas_copy.

Guoguo

On Tue, Jan 5, 2016 at 4:22 PM, Daniel Povey <notifications@github.com>
wrote:

> In src/cudamatrix/cu-sparse-matrix.h
> <https://github.com/kaldi-asr/kaldi/pull/434#discussion_r48914937>:
>
> > @@ -95,6 +95,10 @@ class CuSparseMatrix {
> >    template <typename OtherReal>
> >    void CopyToSmat(SparseMatrix<OtherReal> *smat) const;
> >
> > +  /// Copy elements to CuVector. It is the caller's responsibility to resize
> > +  /// <*vec>.
> > +  void CopyElementsToVec(CuVector<Real> *vec) const;
>
> should be CuVectorBase
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/434/files#r48914937>.
>

danpovey commented:
I'll wait till tests finish before merging.  Thanks.
